core/domain/topic_services.py

```diff
@@ -189,6 +193,185 @@ def save_new_topic(committer_id, topic):
+                    topic.update_skill_ids(change.new_value)
+                elif (change.property_name ==
+                      topic_domain.TOPIC_PROPERTY_LANGUAGE_CODE):
+                    topic.update_language_code(change.new_value)
+        return topic
+
+    except Exception as e:
+        logging.error(
+            '%s %s %s %s' % (
+                e.__class__.__name__, e, topic_id, change_list)
+        )
+        raise
+
+
+def _save_topic(committer_id, topic, commit_message, change_list):
+    """Validates a topic and commits it to persistent storage. If
+    successful, increments the version number of the incoming topic domain
+    object by 1.
+
+    Args:
+        committer_id: str. ID of the given committer.
+        topic: Topic. The topic domain object to be saved.
+        commit_message: str. The commit message.
+        change_list: list(dict). List of changes applied to a topic. Each
+            entry in change_list is a dict that represents a TopicChange.
+
+    Raises:
+        Exception: Received invalid change list.
+        Exception: The topic model and the incoming topic domain
+            object have different version numbers.
+    """
+    if not change_list:
+        raise Exception(
+            'Unexpected error: received an invalid change list when trying to '
+            'save topic %s: %s' % (topic.id, change_list))
+
+    topic_model = topic_models.TopicModel.get(topic.id, strict=False)
+    if topic_model is None:
+        topic_model = topic_models.TopicModel(id=topic.id)
+    else:
+        if topic.version > topic_model.version:
+            raise Exception(
+                'Unexpected error: trying to update version %s of topic '
+                'from version %s. Please reload the page and try again.'
+                % (topic_model.version, topic.version))
+        elif topic.version < topic_model.version:
+            raise Exception(
+                'Trying to update version %s of topic from version %s, '
+                'which is too old. Please reload the page and try again.'
+                % (topic_model.version, topic.version))
+
+    topic_model.description = topic.description
+    topic_model.name = topic.name
+    topic_model.canonical_story_ids = topic.canonical_story_ids
+    topic_model.additional_story_ids = topic.additional_story_ids
+    topic_model.skill_ids = topic.skill_ids
+    topic_model.language_code = topic.language_code
+    topic_model.commit(committer_id, commit_message, change_list)
+    memcache_services.delete(_get_topic_memcache_key(topic.id))
+    topic.version += 1
+
+
+def update_topic(
+        committer_id, topic_id, change_list, commit_message):
+    """Updates a topic. Commits changes.
+
+    Args:
+    - committer_id: str. The id of the user who is performing the update
+        action.
+    - topic_id: str. The topic id.
+    - change_list: list of dicts, each representing a TopicChange object.
```

Comment:
Use domain objects if possible.

Docstring here is formatted differently from the other docstrings.

Reply:
Done.